xdsl/irdl/constraints.py

```diff
@@ -20,38 +21,35 @@ class ConstraintContext:
     Contains the assignment of constraint variables.
     """
 
-    _variables: dict[str, Attribute] = field(default_factory=dict)
+    _variables: ScopedDict[str, Attribute] = field(default_factory=lambda: ScopedDict())
     """The assignment of constraint variables."""
 
-    _range_variables: dict[str, tuple[Attribute, ...]] = field(default_factory=dict)
+    _range_variables: ScopedDict[str, tuple[Attribute, ...]] = field(
+        default_factory=lambda: ScopedDict()
```

Comment:
```suggestion
        default_factory=ScopedDict
```

Reply:
similarly here, Pyright doesn't like it :(